### DISCUSSION TITLE
RigidBody3D sinking

### DISCUSSION THREAD

**Discussion by Mangonels** [godot-jolt/godot-jolt#904] [Questions] [answered]
This might be some sort of oversight on my side, but I'm at my wit's end with it so I decided I'd just open a ticket for orientation on what is going on.

When dropping a rigid body box onto a another collider, it most likely will sink a small bit and resurface to where it should really stay.

This seems to be random and I couldn't identify a pattern to it, the following video proves it at 2:26:

https://www.youtube.com/watch?v=Xi6eTJAO-1A

A rigidbody3d cube is dropped onto a static cube.

In the first setup (identical to the second) the rigidbody stops as soon as it collides with the static body (1:07).

After that I delete the first setup components from the scene and recreate them (resulting in the "second setup", which is basically identical to the first).

In the second setup the rigidbody sinks slightly into the static body (2:26) for whatever reason, and quickly comes out to where it should be.

I've also seen this happening with different collision shapes such as capsule and concave collision shapes.

In all cases the only force applied is the one by the default engine gravity.

Godot 4.2.2, Jolt 0.12.0

Here's a project zip with the test as of when I started reproducing the issue in the video: https://drive.google.com/file/d/1rip5PhvN7zFJ0h9XDxbfmgpzu2zXK1Ho/view?usp=sharing

Note: I just noticed I didn't enable CCD on the rigidbody and the project settings physics gravity is somehow at 25 instead of 9.8, yet I can still reproduce the overall sinking issue in other projects with CCD enabled and a lower gravity such as 9.8.

**Comment by jrouwe** [comment]
This behavior is to be expected. If you're not using CCD then a rigid body will do a full time step at its current velocity + gravity. This can make it end up inside the box a little bit (especially if gravity is much higher than normal). The simulation will detect this the next frame and push the box out again. There is a global setting that indicates how much penetration is allowed under `physics/jolt_3d/solver/contact_allowed_penetration` (default 2 cm). If you want less penetration, lower that value. If you want to further avoid penetration turn on CCD.

**Comment by Mangonels** [comment]
The sinking can occur even with CCD enabled and a very low Contact Allowed Penetration, 

Here's another example on my main project, I made a character controller from a RigidBody3D with a capsule collision shape:

https://youtu.be/TvDZMLPzQUI

The character has CCD active (I show it's active in remote runtime) and I've set Contact Allowed Penetration to 0.001.
Gravity is 100 in the -1 y axis. Default 9.8 makes jumping very floaty for some reason, but I still try it out at 0.40, the rigidbody is shown to still clip at 1:18

**Comment by mihe** [comment]
@jrouwe I'm assuming that the Baumgarte stabilization plays some role in the discrepancy with Godot Physics as well, right?

While the slightly bigger resulting overlap can be explained by `physics/jolt_3d/solver/contact_allowed_penetration`, increasing the Baumgarte stabilization factor (aka `physics/jolt_3d/solver/position_correction`) seems to make things behave more like Godot Physics does, which also presents this behavior (as best I can tell) but obviously resolves the penetration much quicker (presumably at the cost of simulation stability).

**Comment by jrouwe** [comment]
That should indeed resolve the penetration more quickly but can lead to instability as you say.

Preventing penetration in the first place may also be achieved by lowering the `physics/jolt_3d/continuous_cd/movement_threshold` option. A CCD body will only do CCD if it moves more than this percentage of its inner radius*. The idea being that is that if you move less than that, you will penetrate the object but not tunnel through it and the penetration will be fixed in subsequent frames. 

Another option `physics/jolt_3d/continuous_cd/max_penetration` controls how far CCD will move the body into another body if it detects an overlap during CCD. The idea behind this option is that you get a more stable contact if you make the two bodies penetrate a little bit. Again this is a percentage of the inner radius.

I wouldn't set both options to zero, but you can make them much lower than their defaults.

*Inner radius = the radius of the biggest sphere that can be fully contained by the collision volume.

B.t.w. Jolt has not been tuned for a gravity of 100. Various internal tolerances are used that may need revisiting under these gravity conditions.

**Comment by jrouwe** [comment]
One more thing: This whole thing is very much time step dependent. Jolt works best if you take time steps that are smaller or equal to 1/60 s. Your body sinking in the floor may be related to frame rate dips (if the frame delta time is directly passed to physics).

**Comment by mihe** [comment]
> if the frame delta time is directly passed to physics

It's not. Godot uses a fixed physics timestep, which defaults to 60 hz, but is configurable.

**Comment by Mangonels** [comment, marked as answer]
Hello, I think I solved this by just moving `physics/jolt_3d/continuous_cd/movement_threshold` a bit. 

Well, down to exactly 50.2 although I'm sure other values would also do.

Anyway you guys gave me quite a repertoire of useful values to tweak so I hope anyone reading this having same/similar issue can also try them out. (and I also hope my issue doesn't reemerge and I have to come back here crying 😂)
